client-mr/core/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java

```diff
@@ -75,7 +77,11 @@ public Roaring64NavigableMap fetchAllRssTaskIds() {
     String errMsg = "TaskAttemptIDs are inconsistent with map tasks";
     for (TaskAttemptID taskAttemptID : successMaps) {
       if (!obsoleteMaps.contains(taskAttemptID)) {
-        long rssTaskId = RssMRUtils.convertTaskAttemptIdToLong(taskAttemptID, appAttemptId);
+        int maxFailures = jobConf.getInt(MRJobConfig.MAP_MAX_ATTEMPTS, 4);
+        boolean speculation = jobConf.getBoolean(MRJobConfig.MAP_SPECULATIVE, true);
+        int maxAttemptNo = ClientUtils.getMaxAttemptNo(maxFailures, speculation);
+        int rssTaskId =
+            RssMRUtils.createRssTaskAttemptId(taskAttemptID, appAttemptId, maxAttemptNo);
```

Comment:
I'd suggest to move some of this logic into `RssMRUtils`:
```suggestion
        int rssTaskId =
            RssMRUtils.createRssTaskAttemptId(taskAttemptID, appAttemptId, maxFailures, speculation);
```
As a convenience function, you could also provide (but keep above method for testing):
```suggestion
        int rssTaskId =
            RssMRUtils.createRssTaskAttemptId(taskAttemptID, appAttemptId, jobConf);
```